contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,216 @@
+     * @dev Each forwarder will only know if they are on the layer directly before the layer which contains the `target` contract.
+     * If the next layer contains the target, we send `message` cross-chain normally. Otherwise, we wrap `message` into a `relayMessage`
+     * call to the next layer's forwarder contract.
+     */
+    function relayMessage(address target, bytes memory message) external payable override onlyAdmin {
+        RouteIndex storage route = availableRoutes[target];
+        if (route.adapter == address(0)) revert UninitializedRoute();
+        bool success;
+        // Case 1: We are on the network immediately before our target, so we forward the original message over the canonical messaging bridge.
+        if (target == route.spokePool || target == route.forwarder) {
+            (success, ) = route.adapter.delegatecall(abi.encodeCall(AdapterInterface.relayMessage, (target, message)));
+        } else {
+            if (route.forwarder == address(0)) revert MalformedRoute();
+            // Case 2: We are not on the network immediately before our target, so we wrap the message in a `relayMessage` call and forward it
+            // to the next forwarder on the path to the target contract.
+            bytes memory wrappedMessage = abi.encodeCall(AdapterInterface.relayMessage, (target, message));
+            (success, ) = route.adapter.delegatecall(
+                abi.encodeCall(AdapterInterface.relayMessage, (route.forwarder, wrappedMessage))
+            );
+        }
+        if (!success) revert RelayMessageFailed();
+        emit MessageForwarded(target, message);
+    }
+
+    /**
+     * @notice Relays `amount` of a token to `target` over an arbitrary number of layers.
+     * @param baseToken This layer's address of the token to send.
+     * @param remoteToken The next layer's address of the token to send.
+     * @param amount The amount of the token to send.
+     * @param target The address of the contract that will receive the tokens.
+     * @dev The relayTokens function (which must be implemented on a per-chain basis) needs to inductively determine the address of the
+     * remote token given a known baseToken. Then, once the correct addresses are obtained, this function may be called to finish the
+     * relay of tokens.
+     */
+    function _relayTokens(
+        address baseToken,
+        address remoteToken,
+        uint256 amount,
+        address target
+    ) internal {
+        RouteIndex storage route = availableRoutes[target];
+        if (route.adapter == address(0)) revert UninitializedRoute();
+        bool success;
+        // Case 1: We are immediately before the target spoke pool, so we send the tokens to the spoke pool
+        // and do NOT follow it up with a message.
+        if (target == route.spokePool) {
+            (success, ) = route.adapter.delegatecall(
+                abi.encodeCall(AdapterInterface.relayTokens, (baseToken, remoteToken, amount, target))
```

Comment:
If you're going to use `remote` here to reference the "next layer" then that's confusing because you should use the same word to reference possible target contracts in the next layer, but instead you name that list `availableRoutes`. I think you should use the same word to reference either {tokens} or {possible target} contracts on the next layer when naming those `mappings`

Reply:
This (and the comment below) was addressed here https://github.com/across-protocol/contracts/pull/609/commits/d33099343f18b48c9db005412529ced8980b7c86